### PR TITLE
lib/url.py: Strengthen URL escaping.

### PR DESCRIPTION
**Pull request by refeed**
Use the same implementation of how zulip sanitizes its stream name and
URL into a safe URL.

Closes https://github.com/zulip/zulip-archive/issues/35

### CODE REVIEW

**Comment by refeed** [comment]
If anyone asked, I chose the `A-` prefix because `A` for "archive" well it is arbitrary though can be anything but dot char

**Comment by refeed** [comment]
I think this may also closes https://github.com/zulip/zulip-archive/issues/51

**Comment by rht** [comment]
Having a different URL scheme from the way the webapp does it would complicate the mapping from the URLS from logged out view to the ones in this PR.

Since we are not using Jekyll anymore, I think we shouldn't add a case for it either.
The Lean community probably still uses an older version with Jekyll (see https://leanprover-community.github.io/archive/), but I think at this point, it was mainly for the nice UI theming. I'd assume they'd happily switch to the current zulip-archive if it has nice UI.

**Comment by refeed** [comment]
I see, just updated the PR again!

File: lib/url.py [hunk 1]
```diff
@@ -66,24 +66,11 @@ def archive_message_url(
...
+    # It copies zulip's way of encoding stream and topic name to URL
     return (
-        "".join(
-            filter(
-                lambda x: x.isalnum or x == " ",
-                s.encode("ascii", "ignore").decode("utf-8"),
-            )
-        )
-        .replace(" ", "-")
-        .replace("?", "%3F")
```
Comment (rht) [inline lib/url.py:79]: This needs to be investigated. It's due to #34. GitHub Pages might have a stricter set of URL escaping than Zulip Server.

Comment (rht) [reply lib/url.py:79]: What about the topic sanitization? Does it not include " " -> "-" in the webapp?

Comment (refeed) [reply lib/url.py:79]: As far as I see it doesn't

Here's an example, the original topic name is `crab rave`
```
https://chat.zulip.org/#narrow/stream/276-music/topic/crab.20rave
```

Comment (refeed) [reply lib/url.py:79]: This is how the web app makes the topic url:

https://github.com/zulip/zulip/blob/de31114d700561f32139a63a0e5f33d5c30039b3/zerver/lib/url_encoding.py#L41

File: lib/url.py [hunk 2]
```diff
@@ -66,29 +66,28 @@ def archive_message_url(
...
+    """
+    Sanitize the string to a safe string that can be used in URLs
 
-# create a unique sanitized identifier for a topic
-def sanitize_topic(topic_name):
+    This is copied from Zulip's core code:
+    https://github.com/zulip/zulip/blob/de31114d700561f32139a63a0e5f33d5c30039b3/zerver/lib/url_encoding.py#L8
+    """
     return (
-        urllib.parse.quote(topic_name, safe="~()*!.'")
```
Comment (rht) [inline lib/url.py:86]: The "~()*!.'" seems to be arbitrary. It's from this commit: https://github.com/zulip/zulip-archive/commit/f8bea8a9cb55ac318009f07888e6bd510024679e

**Comment by rht** [comment]
LGTM. @timabbott FYI

**Comment by timabbott** [comment]
Merged, thanks @refeed!

**Comment by eric-wieser** [comment]
Isn't this going to break any links into existing archives that use the old encoding scheme?

**Comment by rht** [comment]
That is indeed a problem. The Isabelle Zulip had reported this in the past: https://chat.zulip.org/#narrow/stream/127-integrations/topic/zulip-archive.3A.20FileNotFoundError.

**Comment by eric-wieser** [comment]
Generating redirect HTML pages at all the old URLs would be a reasonable compromise

**Pull request merged**